handler/src/main/java/io/netty/handler/ssl/Java9SslUtils.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl;
+
+import io.netty.util.internal.EmptyArrays;
+import io.netty.util.internal.PlatformDependent;
+
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLParameters;
+import java.lang.reflect.Method;
+import java.security.AccessController;
+import java.security.PrivilegedExceptionAction;
+import java.util.List;
+
+final class Java9SslUtils {
+    private static final Method setApplicationProtocols;
+    private static final Method getApplicationProtocol;
+
+    static {
+        Method setter = null;
+        Method getter = null;
+
+        if (PlatformDependent.javaVersion() >= 9) {
+            try {
+                getter = AccessController.doPrivileged(new PrivilegedExceptionAction<Method>() {
+                    @Override
+                    public Method run() throws Exception {
+                        return SSLEngine.class.getMethod("getApplicationProtocol");
+                    }
+                });
+                setter = AccessController.doPrivileged(new PrivilegedExceptionAction<Method>() {
+                    @Override
+                    public Method run() throws Exception {
+                        return SSLParameters.class.getMethod("setApplicationProtocols", String[].class);
+                    }
+                });
+            } catch (Throwable t) {
+                throw new RuntimeException("Unable to initialize Java9SslUtils", t);
```

Comment:
Instead of throwing a runtime exception inside the static init block we could assign a new static supportsAlpn boolean (including the invoke test) then simply return that in the function below.